### ISSUE TITLE
--observatory-port stopped working in SDK 1.22.3

### ISSUE THREAD

**Issue by wer32** [flutter/flutter#69722]
Hi Everyone,

I used the argument `--observatory-port` for running integration tests locally but in a current version, it stopped working.

```
flutter doctor -v
[✓] Flutter (Channel stable, 1.22.3, on Mac OS X 10.15.7 19H2, locale en-RU)
    • Flutter version 1.22.3 at /Users/sergey.bartosik/Applications/sdk/flutter
    • Framework revision 8874f21e79 (5 days ago), 2020-10-29 14:14:35 -0700
    • Engine revision a1440ca392
    • Dart version 2.10.3

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/sergey.bartosik/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /Users/sergey.bartosik/Library/Android/sdk
    • Java binary at: /Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/bin/java
    • Java version Java(TM) SE Runtime Environment (build 1.8.0_211-b12)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    • CocoaPods version 1.9.3

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).

[✓] IntelliJ IDEA Community Edition (version 2020.2.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin installed
    • Dart plugin version 202.8070

[✓] IntelliJ IDEA Community Edition (version 2020.2.3)
    • IntelliJ at /Applications/ /IntelliJ IDEA CE.app
    • Flutter plugin installed
    • Dart plugin version 202.8070

[✓] Connected device (1 available)
    • iPhone 12 (mobile) • 8FC4CC8F-1509-42C1-A72C-043FC48C2D0C • ios • com.apple.CoreSimulator.SimRuntime.iOS-14-1 (simulator)
```

Run setup SDK 1.22.3
<img width="762" alt="broken logic" src="https://user-images.githubusercontent.com/262981/98051203-f71e8700-1df8-11eb-9493-afa964605cb6.png">

Result for SDK 1.22.3
<img width="1079" alt="run arguments" src="https://user-images.githubusercontent.com/262981/98051260-187f7300-1df9-11eb-86fa-88af00e4400d.png">

However, if I run it for the previous version of SKD 1.20.4 it works perfectly fine
```
flutter doctor -v
[✓] Flutter (Channel unknown, 1.20.4, on Mac OS X 10.15.7 19H2, locale en-RU)
    • Flutter version 1.20.4 at /Users/sergey.bartosik/Applications/sdk/flutter
    • Framework revision fba99f6cf9 (7 weeks ago), 2020-09-14 15:32:52 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/sergey.bartosik/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /Users/sergey.bartosik/Library/Android/sdk
    • Java binary at: /Library/Java/JavaVirtualMachines/jdk1.8.0_211.jdk/Contents/Home/bin/java
    • Java version Java(TM) SE Runtime Environment (build 1.8.0_211-b12)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    • CocoaPods version 1.9.3

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).

[✓] IntelliJ IDEA Community Edition (version 2020.2.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin installed
    • Dart plugin version 202.8070

[✓] IntelliJ IDEA Community Edition (version 2020.2.3)
    • IntelliJ at /Applications/ /IntelliJ IDEA CE.app
    • Flutter plugin installed
    • Dart plugin version 202.8070

 
[✓] Connected device (1 available)            
    • iPhone 12 (mobile) • 8FC4CC8F-1509-42C1-A72C-043FC48C2D0C • ios • com.apple.CoreSimulator.SimRuntime.iOS-14-1 (simulator)

! Doctor found issues in 1 category.
```
<img width="710" alt="correct logic" src="https://user-images.githubusercontent.com/262981/98052228-4960a780-1dfb-11eb-8c68-57db4da1dd36.png">

**Comment by darshankawar** [comment]
@wer32 
I verified on both versions (1.22.3 and 1.20.4), but saw similar observatory ports:

```
1.22.3 : flutter: Observatory listening on http://127.0.0.1:59006/jzgVn5_xGFg=/

1.20.4 : I/flutter (11957): Observatory listening on http://127.0.0.1:46352/MjUgD2KSBoM=/
00:00 +0: Counter App (setUpAll)

VMServiceFlutterDriver: Connecting to Flutter application at http://127.0.0.1:59631/MjUgD2KSBoM=/
```

So not sure if what you are seeing is an expected behavior.

**Comment by wer32** [comment]
HI @darshankawar,
Please note that on the screen above I specified port 8888 and as a result, I got http://127.0.0.1:8888/ws on SDK 1.20.4 that was my goal.
<img width="1079" alt="run arguments" src="https://user-images.githubusercontent.com/262981/98115447-74cfaa80-1e6c-11eb-8867-21720f69dacb.png">

With SDK 1.22.3 it chooses a random port. For me, that looks like broken logic, please correct me if I'm wrong.

My original intention was a convenient development environment that described in this article https://medium.com/flutter-community/hot-reload-for-flutter-integration-tests-e0478b63bd54

**Comment by jason-simmons** [comment]
`--observatory-port` is now deprecated.

It has been replaced by the `--host-vmservice-port` and `--dds-port` flags.

**Comment by wer32** [comment]
@jason-simmons would you help to find instruction or documentation on how to use these parameters, please?
For some reason, my tries were not successful.

**Comment by darshankawar** [comment]
@wer32 
Please see [this](https://github.com/flutter/flutter/issues/56044) open issue where it specifies the same.

**Comment by wer32** [comment]
> @wer32
> Please see [this](https://github.com/flutter/flutter/issues/56044) open issue where it specifies the same.

@darshankawar I'm sorry I think it is a different proposal and issue.
As I mentioned above nether --observatory-port nor --host-vmservice-port works properly. I see a random port instead of port 8888. Please see my examples above.

**Comment by jason-simmons** [comment]
Use the `--dds-port` flag to control the port of the observatory URL

**Comment by Kurogoma4D** [comment]
Same here( #69871 ), but I have tried `--dds-port` flag on Flutter beta channel, and it works fine.
That flag is able to use on beta channel, not stable channel.

```
$ flutter --version                                        

Flutter 1.23.0-18.1.pre • channel beta • https://github.com/flutter/flutter.git
Framework • revision 198df796aa (3 weeks ago) • 2020-10-15 12:04:33 -0700
Engine • revision 1d12d82d9c
Tools • Dart 2.11.0 (build 2.11.0-213.1.beta)

$ flutter run --dds-port 8888 --disable-service-auth-codes

Launching lib/main.dart on iPhone 8 Plus in debug mode...
 
Running Xcode build...                                                  
 └─Compiling, linking and signing...                         4.6s
Xcode build done.                                            9.9s
Waiting for iPhone 8 Plus to report its views...                     5ms
Syncing files to device iPhone 8 Plus...                            49ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h Repeat this help message.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).
An Observatory debugger and profiler on iPhone 8 Plus is available at: http://127.0.0.1:8888/

```

But even if this is done, there is still the problem of not being able to do integration testing in a stable channel...

**Comment by jonahwilliams** [comment]
This is a regression. The introduction of dds should not have stopped observatory port from working, but I think this was lost in the weeds when dds was switched from a wrapper to an interceptor? I'm not quite sure.

FYI @bkonyi

**Comment by jonahwilliams** [comment]
I cannot repro the regression in 1.22.3 with the flutter-tester device - its possible that there is an issue with the fallback strategy with mDNS/fallback strategy should be removed now. FYI @jmagman

**Comment by bkonyi** [comment]
@jonahwilliams this is WAI. `--observatory-port` is setting the port for the VM service, not DDS, and we need to be able to set them both separately to handle workflows over SSH. `--dds-port` should be used to control the port displayed by the tool and we should really encourage users to not use the `--observatory-port` option.

**Comment by jonahwilliams** [comment]
Practically everyone that is using --host-vmservice-port or --observatory-port will have there workflows broken though. Can we adjust the behavior there so it is less surprising?

**Comment by bkonyi** [comment]
We'd have to change the meaning of the options and then introduce a new option to allow for setting the VM service port itself. It's unfortunate, but we do need to be able to set both of these ports manually in some scenarios. This workflow change has already been propagated through google3 as well, so it'll be another workflow breakage if we do decide to change this.

**Comment by jonahwilliams** [comment]
I thought originally there was some plan to make them all copacetic, but I could be misremembering.

Can we separate out the google WFH workflow and the external workflow out a bit here? Either that or we need some breaking change comms

**Comment by jmagman** [comment]
Bisected the stable issue to #63416, this line is missing a trailing comma:
https://github.com/flutter/flutter/pull/63416/files#diff-fbf1a86439c67e75e90a00f32fd2b789323582ced2f4f5cfea2d27a1af18aa30R427

The comma was added in #64671.

Will prep a 1.22 hotfix while we decide what to do about `--dds-port` and friends in the next stable.

**Comment by jonahwilliams** [comment]
For --dds-port, how about something like:

If both --host-vmservice-port and --dds-port are specified, use the current behavior.
if only --host-vmservice-port is specified and dds is enabled, use that for the dds port. Otherwise it acts as normal
if only --dds-port is specified, use that for the dds port.

**Comment by bkonyi** [comment]
Sure, that's an option, but that's also not terribly obvious behavior. If you think that's the least confusing behavior to implement then SGTM.

**Comment by jonahwilliams** [comment]
That gives us a chance to go through a deprecation process for the existing arguments. I agree its not the best end state

**Comment by bkonyi** [comment]
Sure, sounds good.

**Comment by jonahwilliams** [comment]
Do you have time to take a look at this in the next week or so?

**Comment by jmagman** [comment]
Cherry-pick request is for https://github.com/flutter/flutter/pull/70325 and the originator's report of this failing on stable.

@bkonyi @jonahwilliams Let's take the wider dds flag discussion to https://github.com/flutter/flutter/issues/70332.

**Comment by jmagman** [comment]
Fixed with https://github.com/flutter/flutter/pull/70325, available on 1.22.4 stable.
